pennylane/beta/tapes/tape.py

```diff
@@ -0,0 +1,776 @@
+                An operation or tape appearing in this list of names is not expanded.
+
+        **Example**
+
+        Consider the following nested tape:
+
+        .. code-block:: python
+
+            with QuantumTape() as tape:
+                qml.BasisState(np.array([1, 1]), wires=[0, 'a'])
+
+                with QuantumTape() as tape2:
+                    qml.Rot(0.543, 0.1, 0.4, wires=0)
+
+                qml.CNOT(wires=[0, 'a'])
+                qml.RY(0.2, wires='a')
+                probs(wires=0), probs(wires='a')
+
+        The nested structure is preserved:
+
+        >>> tape.operations
+        [BasisState(array([1, 1]), wires=[0, 'a']),
+         <QuantumTape: params=3, wires=<Wires = [0]>>,
+         CNOT(wires=[0, 'a']),
+         RY(0.2, wires=['a'])]
+
+        Calling ``.expand`` will return a tape with all nested tapes
+        expanded, resulting in a single tape of quantum operations:
+
+        >>> new_tape = tape.expand()
+        >>> new_tape.operations
+        [PauliX(wires=[0]),
+         PauliX(wires=['a']),
+         Rot(0.543, 0.1, 0.4, wires=[0]),
+         CNOT(wires=[0, 'a']),
+         RY(0.2, wires=['a'])]
+        """
+        old_tape = self
+
+        for _ in range(depth):
+            new_tape = old_tape._expand(stop_at=stop_at)
+
+            if len(new_tape.operations) == len(old_tape.operations) and [
+                o1.name == o2.name for o1, o2 in zip(new_tape.operations, old_tape.operations)
+            ]:
+                # expansion has found a fixed point
+                break
+
+            old_tape = new_tape
+
+        new_tape._update()
+        return new_tape
+
+    def inv(self):
+        """Inverts the processed operations.
+
+        Inversion is performed in-place.
+
+        .. note::
+
+            This method only inverts the quantum operations/unitary recorded
+            by the quantum tape; state preprations and measurements are left unchanged.
+
+        **Example**
+
+        .. code-block:: python
+
+            with QuantumTape() as tape:
+                qml.BasisState(np.array([1, 1]), wires=[0, 'a'])
+                qml.RX(0.432, wires=0)
+                qml.Rot(0.543, 0.1, 0.4, wires=0).inv()
+                qml.CNOT(wires=[0, 'a'])
+                probs(wires=0), probs(wires='a')
+
+        This tape has the following properties:
+
+        >>> tape.operations
+        [BasisState(array([1, 1]), wires=[0, 'a']),
+         RX(0.432, wires=[0]),
+         Rot.inv(0.543, 0.1, 0.4, wires=[0]),
+         CNOT(wires=[0, 'a'])]
+        >>> tape.get_parameters()
+        [array([1, 1]), 0.432, 0.543, 0.1, 0.4]
+
+        Here, lets set some trainable parameters:
+
+        >>> tape.trainable_params = {1, 2}
+        >>> tape.get_parameters()
+        [0.432, 0.543]
+
+        Inverting the tape:
+
+        >>> tape.inv()
+        >>> tape.operations
+        [BasisState(array([1, 1]), wires=[0, 'a']),
+         CNOT.inv(wires=[0, 'a']),
```

Comment:
Not really. We'll need to modify the operations directly to fix this though (feel out of scope for this PR)

Reply:
Would be nice to find a generic solution indeed, so all devices automatically support inversions that do not imply changing the gate itself!

Reply:
Agree! This could be as simple as adding an attribute to unitary operations which are there own inverses.

Then, in `Operation.inv`:

```python
def inv(self):
    if not self.self_inverse:
        current_inv = QueuingContext.get_info(self).get("inverse", False)
        QueuingContext.update_info(self, inverse=not current_inv)
    return self
```